gcc/rust/lex/rust-lex.cc

```diff
@@ -2839,3 +2543,62 @@ Lexer::start_line (int current_line, int current_column)
 }
 
 } // namespace Rust
+
+#if CHECKING_P
+
+namespace selftest {
+
+// Checks if `src` has the same contents as the given characters
+void
+assert_source_content (Rust::Lexer::InputSource &src,
+		       std::vector<uint32_t> expected)
+{
+  Rust::Codepoint src_char = src.next ();
+  for (auto expected_char : expected)
+    {
+      // Make sure that `src` is not shorter than `expected`
+      ASSERT_FALSE (src_char.is_eof ());
+      // Checks skipped character is expeceted one.
+      ASSERT_EQ (src_char.value, expected_char);
+      src_char = src.next ();
+    }
+  // Checks if `src` and `chars` has the same length.
+  ASSERT_TRUE (src_char.is_eof ());
+}
+
+void
+test_buffer_input_source (std::string str, std::vector<uint32_t> expected)
+{
+  Rust::Lexer::BufferInputSource source (str, 0);
+  std::cout << source.is_valid () << std::endl;
+  assert_source_content (source, expected);
+}
+
+void
+rust_input_source_test ()
+{
+  std::string src = u8"_abcde\tXYZ\v\f";
+  std::vector<uint32_t> expected
+    = {'_', 'a', 'b', 'c', 'd', 'e', '\t', 'X', 'Y', 'Z', '\v', '\f'};
+  test_buffer_input_source (src, expected);
```

Comment:
I have no idea how to convert(?) `std::string` into `FILE` so only BufferInputSource is tested now.

Reply:
Added unit tests for `BufferInputSource`.